File: ml-benchmark/synduce/list/sndmin.f
Inductive List = two {Int, Int} | cons {Int, List};
Inductive CList = ctwo {Int, Int} | concat {CList, CList};

max = \a: Int. \b: Int. if < a b then b else a;
min = \a: Int. \b: Int. if > a b then b else a;

cat_list = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    two {a, b} -> cons {a, cons {b, ys}}
  | cons {hd, tl} -> cons {hd, f tl ys}
  end
);

repr = fix (
  \f: CList -> Compress List. \xs: CList.
  match xs with
    ctwo {a, b} -> two {a, b}
  | concat {a, b} -> cat_list (f a) (f b)
  end
);

spec = \xs: List. (fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    two {a, b} -> {min a b, max a b}
  | cons {hd, tl} ->
    let result = f tl in
    let new_min = min hd result.1 in
    let new_snd = min result.2 (max hd result.1) in
    {new_min, new_snd}
    end
) xs).2;

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/sndmin.ml
type ilist = Two of int * int | Cons of int * ilist
type clist = Ctwo of int * int | Concat of clist * clist

let max a b = if a < b then b else a
let min a b = if a > b then b else a

let rec cat_list xs ys =
  match xs with
  | Two (a, b) -> Cons (a, Cons (b, ys))
  | Cons (hd, tl) -> Cons (hd, cat_list tl ys)

let rec repr xs =
  match xs with
  | Ctwo (a, b) -> Two (a, b)
  | Concat (a, b) -> cat_list (repr a) (repr b)

let spec xs =
  let rec f xs =
    match xs with
    | Two (a, b) -> (min a b, max a b)
    | Cons (hd, tl) ->
        let result = f tl in
        match result with
        | (r1, r2) ->
            let new_min = min hd r1 in
            let new_snd = min r2 (max hd r1) in
            (new_min, new_snd)
  in
  match f xs with
  | (_, s2) -> s2

let program x = spec (repr x)